src/client/kotlin/me/owdding/skyocean/features/dev/NpcRecipeParser.kt

```diff
@@ -0,0 +1,142 @@
+    val clickToRegex = Regex("[\\s\\S\\n]+Click to \\S+!")
+    val amountRegex = Regex(".* (x[\\d,.]+)")
+
+    @OptIn(ExperimentalContracts::class)
+    private inline fun <T, R> T.ifEnabled(block: T.() -> R) {
+        contract {
+            callsInPlace(block, InvocationKind.AT_MOST_ONCE)
+        }
+
+        if (!enable) return
+        block()
+    }
+
+    @Subscription
+    @OnlyOnSkyBlock
+    @MustBeContainer
+    private fun InventoryChangeEvent.inventory() = ifEnabled {
+        if (isSkyBlockFiller) return
+        val lore = item.getRawLore().joinToString("\n")
+        if (!lore.trim().matches(clickToRegex)) return
+
+        val costs = lore.substringAfter("Cost").split("\n").takeWhile { it.isNotBlank() }.map {
+            runCatching { parseIngredient(it) }.getOrElse {
+                it.printStackTrace()
+                null
+            }
+        }
+        if (costs.isEmpty()) return
+
+        val (_, amount) = item.cleanName.getAmount()
+        val output = runCatching { item.toIngredient(amount) }.getOrElse {
+            it.printStackTrace()
+            null
+        }
+        lastInv[slot.index] = MaybeCustomRecipe(CustomRecipe(output, costs.filterNotNull().toMutableList()), false)
+    }
+
+    fun String.getAmount(): Pair<String, Int> = if (this.matches(amountRegex)) {
+        val amount = this.replace(amountRegex, "$1")
+        this.removeSuffix(amount).trim() to amount.substring(1).toIntValue()
+    } else {
+        this to 1
+    }
+
+    private fun parseIngredient(cost: String): Ingredient? = when {
+        cost.endsWith("Coins") -> CurrencyIngredient.coins(cost.filter { it.isDigit() }.toIntValue())
+        cost.endsWith("Chocolate") -> CurrencyIngredient(
+            cost.filter { it.isDigit() }.toIntValue(),
+            CurrencyType.CHOCOLATE
+        )
+
+        else -> {
+            val (cost, amount) = cost.getAmount()
+
+            //val itemByName = SimpleItemApi.findIdByName(cost)?.toItem()
+
+            //itemByName.toIngredient(amount)
+            null
+        }
+    }
+
+    private fun ItemStack.toIngredient(amount: Int): ItemLikeIngredient {
+        return TODO()
+    }
+
+    @Subscription
+    @OnlyOnSkyBlock
+    private fun ContainerCloseEvent.onClose() = ifEnabled {
+        var saved = 0
+        lastInv.values.filter { it.save }.map { it.recipe }.forEach {
+            path.createDirectories()
+            val recipe = path.resolve("${it.output?.skyblockId}.json")
+            recipe.writeText(
+                it.toJson(SkyOceanCodecs.CustomRecipeCodec.codec()).toPrettyString(),
+                Charsets.UTF_8,
+                StandardOpenOption.CREATE,
+                StandardOpenOption.TRUNCATE_EXISTING,
+            )
+            saved++
+        }
+        Text.of("Saved $saved recipes").sendWithPrefix()
+        lastInv.clear()
+    }
+
+    @Subscription
+    @OnlyOnSkyBlock
+    private fun ScreenMouseClickEvent.Pre() = ifEnabled {
```

Comment:
Is this correct? because this just means you are looking for both pre and post and your function name is `Pre`